USB: rework USB descriptors (RFC)

jfischer-no opened the pull request:
This PR is the first step towards modular USB class drivers and functions.

- add a RAM section for the USB descriptors
- add USB descriptor section macros
- move class descriptors to the corresponding class driver

Tested on FRDM K64F and Arduino 101

jfischer-no committed:
include: linker: add sections for USB device stack

This patch adds a RAM section for the USB descriptors and
USB interfaces data.

The parts of USB descriptor can be declared independently in
different USB class drivers or functions. These must then be
placed in a predefined order one after the other without gaps
in a RAM section. There may also be several device descriptors
in the section such as for USB DFU.

Since at runtime the number of interfaces and endpoints in the
composite device descriptor is unknown, they must be corrected.
The class drivers or functions use the usb_cfg_data struct to
store the callbacks, interface and endpoints data. This struct
must also be corrected and needs to be placed in the data section.
The descriptor section and data section must be processed by
the USB Device Stack before the enumeration.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
include: usb: add descriptor and data section macros

This patch adds USB descriptor and data section macros.

Use USBD_DESCR_..._DEFINE macros to place the parts of USB device
descriptor in predetermined order in the USB descriptor
section. The parts of a device descriptor are also
sorted according to the macro argument. All parts of a
particular device descriptor must use the same argument.

Use USBD_CFG_DATA_DEFINE macro to place the data struct
of a class driver or function in the USB data section.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented:
@loicpoulain please review (could not insert you as a reviewer)

finikorg reviewed (dismissed):
For me even sanitycheck fails:
sanitycheck -t usb
...
10 tests selected, 31165 tests discarded due to filters
total complete:    1/  10  10%  failed:    0

arduino_101               samples/subsys/usb/mass/test                       FAILED: build_error
	see: ../../../sanity-out/arduino_101/samples/subsys/usb/mass/test/build.log

total complete:    5/  10  50%  failed:    1

quark_se_c1000_devboard   samples/net/http_server/test_usbnet                FAILED: build_error
	see: ../../../sanity-out/quark_se_c1000_devboard/samples/net/http_server/test_usbnet/build.log

total complete:    8/  10  80%  failed:    2

quark_se_c1000_devboard   samples/net/echo_server/test_usbnet_composite      FAILED: build_error
	see: ../../../sanity-out/quark_se_c1000_devboard/samples/net/echo_server/test_usbnet_composite/build.log

total complete:    9/  10  90%  failed:    3

quark_se_c1000_devboard   samples/net/echo_server/test_usbnet                FAILED: build_error
	see: ../../../sanity-out/quark_se_c1000_devboard/samples/net/echo_server/test_usbnet/build.log

total complete:   10/  10  100%  failed:    4

jfischer-phytec-iot commented:
@finikorg I wanted to have feedback on the proposal first, I will fix and test netusb next.

finikorg commented:
@jfischer-phytec-iot the idea is good, the issue which we discussed last time remains: structures redefinitions in several places (at least for CDC), they should be merged at some point.

jfischer-phytec-iot commented:
>tructures redefinitions in several places (at least for CDC), they should be merged at some point.

@finikorg there was the wrong header in netusb, cdc_acm.h is outdated and will be removed after webusb is reworked.

jfischer-phytec-iot commented:
@nashif, sanitycheck fails with: `FAILED: quark_se_c1000_devboard/samples/subsys/usb/dfu/test_x86 has unrecognized binary sections: ['usb_descriptor']`, any idea how to make new section _recognized_?

finikorg commented:
Tested, works with quark_se_c1000_devboard. Now because of this stupid github interface I do not know how to cancel my previous change request and approve this PR. @nashif do you know how to do it?

jfischer-phytec-iot commented:
Attention it is still RFC/WIP, 7ccbd56 will be reworked to support secondary device descriptor (USB DFU needs it).

IvanSanchez commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -75,6 +75,130 @@ static struct uart_driver_api cdc_acm_driver_api;
+		.bNumEndpoints = 1,
+		.bInterfaceClass = COMMUNICATION_DEVICE_CLASS,
+		.bInterfaceSubClass = ACM_SUBCLASS,
+		.bInterfaceProtocol = V25TER_PROTOCOL,
```
Please set this to zero. USB CDC ACM interfaces should advertise support for the AT V2.50 protocol only if they actually implement it. Otherwise, it [will cause problems](https://bugs.freedesktop.org/show_bug.cgi?id=85007).

carlescufi commented:
@jfischer-phytec-iot now that Nordic (almost) has a USB driver as well we're interested in pushing this forward, is there a way we can help?

jfischer-phytec-iot commented:
@carlescufi This is outdated, I need 2-3 days to update and rebase it, meanwhile I would be happy about review, help and feedback for #4373

jfischer-no committed:
subsys: usb: add function to find and fix USB descriptors

This patch introduces a function to find and correct all string
descriptor.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
subsys: usb: move descriptor parts to the class drivers

This patch moves the descriptor parts to related class
and function drivers and extends the usb_fix_descriptor
function so that the wTotalLength and bNumInterfaces
are corrected before enumeration.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
subsys: usb: rework usb string descriptor fixup

This patch simplifies the handling of the string descriptors.
It introduces common macro for all string descriptors to
calculate the length of the bString.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
include: driver: usb: add check for endpoint capabilities

Add API function to check capabilities of an endpoint.
It should be used to check the capabilities of an endpoint
before configuration. The driver should return 0 if the
endpoint configuration is possible.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
driver: usb: add check for endpoint capabilities

Add function to check capabilities of an endpoint.
Only basic properties are checked, especially on STM32
capabilities of different USB controller configurations
have to be considered in the future.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
subsys: usb: validate and update endpoint address

Add the routine to validate endpoint address and update the endpoint
descriptors and usb_ep_cfg_data at runtime.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
subsys: usb: fetch endpoint address from usb_ep_cfg_data

This patch changes all class functions so that the endpoint
address will be retrieved from the usb_ep_cfg_data at runtime.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
subsys: usb: configure Interface descriptor at runtime

Introduce function to configure interface descriptor at runtime.
It is simple to leave the corresponding function to configure the
interface descriptor and fix bInterfaceNumber and iInterface values,
for example.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

carlescufi commented:
@jfischer-phytec-iot are you going to rework this now that the DFU PR is merged?

carlescufi commented:
@finikorg @IvanSanchez is this ready to be merged?

carlescufi commented:
@sathishkuttan can you take a look at this PR please?

carlescufi commented:
@jfischer-phytec-iot does this need to update `webusb.c` as well?

finikorg commented on subsys/usb/usb_descriptor.c:
```diff
@@ -826,41 +172,91 @@ static void ascii7_to_utf16le(void *descriptor)
+		return NULL;
+	}
+
+	return (u8_t *) &common_desc;
```
I think you could use here also __usb_descriptor_start

finikorg reviewed (commented):
Generally LGTM with some comments above inline ones:
- are you able to run sanitycheck -t usb ?
I get errors like:
24 tests selected, 53639 tests discarded due to filters
total complete:   24/  24  100%  failed:    0
FAILED: 96b_carbon/samples/net/echo_server/test_usbnet has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: 96b_carbon/samples/net/echo_server/test_usbnet_composite has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: 96b_carbon/samples/net/http_server/test_usbnet_ecm has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: 96b_carbon/samples/net/http_server/test_usbnet_eem has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: 96b_carbon/samples/net/zperf/test_netusb_ecm has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: 96b_carbon/samples/net/zperf/test_netusb_eem has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: 96b_carbon/samples/net/zperf/test_netusb_rndis has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: arduino_101/samples/subsys/usb/console/test has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: arduino_101/samples/subsys/usb/mass/test has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: nrf52840_pca10056/samples/bluetooth/hci_usb/usb.bluetooth has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: nrf52840_pca10056/samples/subsys/usb/dfu/test has unrecognized binary sections: ['usb_descriptor']
FAILED: quark_se_c1000_devboard/samples/bluetooth/hci_usb/usb.bluetooth has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/net/echo_server/test_usbnet has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/net/echo_server/test_usbnet_composite has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/net/http_server/test_usbnet_ecm has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/net/http_server/test_usbnet_eem has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/net/wpan_serial/test_x86 has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/net/zperf/test_netusb_ecm has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/net/zperf/test_netusb_eem has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/net/zperf/test_netusb_rndis has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/subsys/usb/hid/test_hid has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/subsys/usb/hid/test_hid_composite has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/subsys/usb/hid/test_hid_msc_composite has unrecognized binary sections: ['usb_descriptor', 'usb_data']
FAILED: quark_se_c1000_devboard/samples/subsys/usb/webusb/test_x86 has unrecognized binary sections: ['usb_descriptor']

sathishkuttan reviewed (changes requested):
With use of `QM_` definitions, the build was not successful for me.

jfischer-no committed:
subsys: usb: rework composite device support

This patch reworks composite device support. It allows the
functions to be modular and the user to combine its own
functions with the USB functions of the Zephyr OS.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
usb: class: adapt functions for new composite interface

Adapt functions for new composite interface.

Assign bInterfaceNumber and similar variables of a Interface
default values, these should be valid values for non-composite
configuration. For the case of the composite configuration,
these variables must be set by the interface configuration
function (interface_config) accordingly.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-phytec-iot commented:
Two recent commits is actually the reason for this PR. The respective functions are now modular and something like `usb_descriptor.h` does not have to be touched every time a new function is inserted. Also the automatic configuration of the endpoints is present, though not perfect. (I've just noticed that CONFIG_s for the endpoint addresses are still there and can be removed, also the usb_ep_cfg_data can be tweaked because some information is duplicated, but I do not want to do it in the PR, otherwise it ends up in a rebase hell.)

I did not test everything with all supported controllers, mostly with frdm_k64f. RNDIS and HID have bugs in composite configuration. Other combinations should be ok, tested: MSD + ECM, MSD + EEM, MSD + ACM.

carlescufi commented:
@jfischer-phytec-iot I am not sure I follow. So this PR is now ready to be merged with the addition of the 2 last commits?

jfischer-phytec-iot commented:
@carlescufi yes, if it is considered acceptable.

finikorg committed:
usb: tests: Add missing sections to sanitycheck

Fix warnings during sanitycheck

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg reviewed (approved):
LGTM, with my fix to sanitycheck.

nashif commented:
rebase please

jfischer-phytec-iot commented:
rebased, went through rebase hell, hope nothing is broken

jfischer-phytec-iot commented:
recheck

sathishkuttan commented on drivers/usb/device/usb_dc_dw.c:
```diff
@@ -814,6 +814,38 @@ int usb_dc_set_address(const u8_t addr)
+	}
+
+	if ((USB_DW_EP_ADDR2DIR(cfg->ep_addr) == USB_EP_DIR_OUT) &&
+	    (ep_idx >= QM_USB_OUT_EP_NUM)) {
```
I am getting build errors when I try to build with latest master `0824ec6`

`
drivers/usb/device/usb_dc_dw.c: In function ‘usb_dc_ep_check_cap’:
drivers/usb/device/usb_dc_dw.c:829: error: ‘QM_USB_MAX_PACKET_SIZE’ undeclared (first use in this function)
drivers/usb/device/usb_dc_dw.c:829: error: (Each undeclared identifier is reported only once
drivers/usb/device/usb_dc_dw.c:829: error: for each function it appears in.)
drivers/usb/device/usb_dc_dw.c:835: error: ‘QM_USB_OUT_EP_NUM’ undeclared (first use in this function)
drivers/usb/device/usb_dc_dw.c:841: error: ‘QM_USB_IN_EP_NUM’ undeclared (first use in this function)
`

jfischer-phytec-iot replied:
Which SoC and board?

sathishkuttan replied:
Intel S1000.
But the point is that the QM_ definitions are all the way in `ext/hal/qmsi/soc/quark_se/include/qm_soc_regs.h`. Why use those definitions in a USB device driver? I had made a suggestion earlier to define one in the USB driver separate from QMSI as the device driver is not only for Quark SoCs

jfischer-phytec-iot replied:
I get your point now. Since the driver is used by several platforms I suggest to use device tree for the configuration and capabilities. Also usb_dw_clock_enable() and usb_dw_clock_disable() should also be revised.

sathishkuttan replied:
I agree with using device tree. Will you p;lease address this in a PR?

sathishkuttan commented on subsys/usb/class/hid/core.c:
```diff
@@ -10,15 +10,64 @@
+
+static void usb_set_hid_report_size(u16_t report_desc_size)
+{
+	hid_cfg.if0_hid.subdesc[0].wDescriptorLength =
```
@jfischer-phytec-iot 
Johann, this assignment was causing an unaligned store exception when I ran the rebased code on Intel S1000. I think this would happen on other architectures that require aligned accesses.
`wDescriptorLength` is a `u16_t`.
The descriptors are arranged in memory such that
The device descriptor is first and occupies 18 bytes
The configuration descriptor is next and occupies 9 bytes
The interface descriptor occupies 9 bytes
The HID class descriptor starts at a relative offset of 18+9+9 = 36 bytes
The `usb_hid_class_subdescriptor` starts at an offset of 36 + 6 = 42 bytes
Within `usb_hid_class_subdescriptor`, the first field is a 1 byte `bDescriptorType` and the 2nd field is a 2 byte `wDescriptorLength` and this 2 byte field starts at an odd address/offset.
And this causes an unaligned data store exception.
I had to use byte assignments temporarily to get around this exception, and I wanted to bring this to your attention to see if you expected this when you reworked the code.

jfischer-phytec-iot replied:
@sathishkuttan Why should not it happen before this PR?

can you please test this:
```
--- a/subsys/usb/class/hid/core.c
+++ b/subsys/usb/class/hid/core.c
@@ -68,8 +68,8 @@ USBD_CLASS_DESCR_DEFINE(primary) struct usb_hid_config hid_cfg = {
 
 static void usb_set_hid_report_size(u16_t report_desc_size)
 {
-       hid_cfg.if0_hid.subdesc[0].wDescriptorLength =
-               sys_cpu_to_le16(report_desc_size);
+       UNALIGNED_PUT(sys_cpu_to_le16(report_desc_size),
+                     &hid_cfg.if0_hid.subdesc[0].wDescriptorLength);
 }
 
 static struct hid_device_info {
```
I suspected it will be necessary at several places.

sathishkuttan replied:
@jfischer-phytec-iot 
Good question. I just tested with the 1.12 release tag and surprisingly, the wDescriptorLength is also at an odd address but there is no exception!
Looks like something else changed in another commit and this is not due to your commit in this PR.
